Public/Src/Tools/DropDaemon/DropDaemon.cs

```diff
@@ -0,0 +1,875 @@
+                {
+                    var dropEvent = Activator.CreateInstance<TDropEvent>();
+                    dropEvent.Succeeded = false;
+                    dropEvent.ErrorMessage = errorMessage;
+                    return dropEvent;
+                });
+        }
+
+        private static async Task<TResult> HandleKnownErrors<TResult>(Func<Task<TResult>> factory, Func<string, TResult> errorValueFactory)
+        {
+            try
+            {
+                return await factory();
+            }
+            catch (VssUnauthorizedException e)
+            {
+                return errorValueFactory("[DROP AUTH ERROR] " + e.Message);
+            }
+            catch (DropServiceException e)
+            {
+                return errorValueFactory("[DROP SERVICE ERROR] " + e.Message);
+            }
+            catch (DaemonException e)
+            {
+                return errorValueFactory("[DAEMON ERROR] " + e.Message);
+            }
+        }
+
+        private static string DropItemToString(DropItem dropItem)
+        {
+            try
+            {
+                return dropItem?.ToJson().ToString();
+            }
+#pragma warning disable ERP022 // TODO: This should really handle specific errors
+            catch
+            {
+                return null;
+            }
+#pragma warning restore ERP022 // Unobserved exception in generic exception handler
+        }
+
+        private static int ComputeDropItemExpiration(DropItem dropItem)
+        {
+            DateTime? expirationDate;
+            return dropItem.TryGetExpirationTime(out expirationDate) || expirationDate.HasValue
+                ? (int)expirationDate.Value.Subtract(DateTime.UtcNow).TotalDays
+                : -1;
+        }
+
+        private async Task<T> SendDropEtwEvent<T>(Task<T> task) where T : DropOperationBaseEvent
+        {
+            long startTime = DateTime.UtcNow.Ticks;
+            T dropEvent = null;
+            try
+            {
+                dropEvent = await task;
+                return dropEvent;
+            }
+            finally
+            {
+                // if 'task' failed, create an event indicating an error
+                if (dropEvent == null)
+                {
+                    dropEvent = Activator.CreateInstance<T>();
+                    dropEvent.Succeeded = false;
+                    dropEvent.ErrorMessage = "internal error";
+                }
+
+                // common properties: execution time, drop type, drop url
+                dropEvent.ElapsedTimeTicks = DateTime.UtcNow.Ticks - startTime;
+                dropEvent.DropType = "VsoDrop";
+                if (m_dropClientTask.IsCompleted && !m_dropClientTask.IsFaulted)
+                {
+                    dropEvent.DropUrl = (await m_dropClientTask).DropUrl;
+                }
+
+                // send event
+                m_etwLogger.Log(dropEvent);
+            }
+        }
+
+        #region Config options and commands

```

Comment:
>#region Config options and commands [](start = 8, length = 35)

Is it worth splitting into different files?

Reply:
Due to how things (options/commands) are interconnected, options/commands should stay in one place. Moving them into a separate (sub)class won't work either (well, it will look even uglier).

---
In reply to: [298344104](https://github.com/microsoft/BuildXL/pull/528#discussion_r298344104) [](ancestors = 298344104)